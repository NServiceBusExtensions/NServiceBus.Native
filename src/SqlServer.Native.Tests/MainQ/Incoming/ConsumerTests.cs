using System.Collections.Concurrent;
using System.Linq;
using System.Threading.Tasks;
using NServiceBus.Transport.SqlServerNative;
using ObjectApproval;
using Xunit;
using Xunit.Abstractions;

public class ConsumerTests :
    TestBase
{
    string table = "ConsumerTests";

    [Fact]
    public async Task Single()
    {
        await TestDataBuilder.SendData(table);
        var consumer = new QueueManager(table, SqlConnection);
        using (var result = consumer.Consume().Result)
        {
            ObjectApprover.VerifyWithJson(result.ToVerifyTarget());
        }
    }

    [Fact]
    public async Task Single_nulls()
    {
        await TestDataBuilder.SendNullData(table);
        var consumer = new QueueManager(table, SqlConnection);
        using (var result = consumer.Consume().Result)
        {
            ObjectApprover.VerifyWithJson(result.ToVerifyTarget());
        }
    }

    [Fact]
    public async Task Batch()
    {
        await TestDataBuilder.SendMultipleDataAsync(table);

        var consumer = new QueueManager(table, SqlConnection);
        var messages = new ConcurrentBag<IncomingVerifyTarget>();
        var result = consumer.Consume(
                size: 3,
                action: message => { messages.Add(message.ToVerifyTarget()); })
            .Result;
        Assert.Equal(3, result.Count);
    }

    [Fact]
    public async Task Batch_all()
    {
        await TestDataBuilder.SendMultipleDataAsync(table);

        var consumer = new QueueManager(table, SqlConnection);
        var messages = new ConcurrentBag<IncomingVerifyTarget>();
        var result = consumer.Consume(
                size: 10,
                action: message => { messages.Add(message.ToVerifyTarget()); })
            .Result;
        Assert.Equal(5, result.Count);
        ObjectApprover.VerifyWithJson(messages.OrderBy(x => x.Id));
    }

    public ConsumerTests(ITestOutputHelper output) : base(output)
    {
        var manager = new QueueManager(table, SqlConnection);
        manager.Drop().Await();
        manager.Create().Await();
    }
}