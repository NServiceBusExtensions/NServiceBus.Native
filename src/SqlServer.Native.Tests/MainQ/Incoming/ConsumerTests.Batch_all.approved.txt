[
  {
    "$type": "IncomingVerifyTarget",
    "Expires": "DateTime 1",
    "Headers": "headers",
    "Id": "Guid 1",
    "Body": "{}"
  },
  {
    "$type": "IncomingVerifyTarget",
    "Headers": "{}",
    "Id": "Guid 2"
  },
  {
    "$type": "IncomingVerifyTarget",
    "Expires": "DateTime 1",
    "Headers": "headers",
    "Id": "Guid 3",
    "Body": "{}"
  },
  {
    "$type": "IncomingVerifyTarget",
    "Headers": "{}",
    "Id": "Guid 4"
  },
  {
    "$type": "IncomingVerifyTarget",
    "Expires": "DateTime 1",
    "Headers": "headers",
    "Id": "Guid 5",
    "Body": "{}"
  }
]